### PR TITLE
x86: Implement msplats opcode

### PR DESCRIPTION
**Pull request by BradleyWood**

### CODE REVIEW

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 1]
```diff
@@ -6030,9 +6030,160 @@ TR::Register *OMR::X86::TreeEvaluator::mstoreiEvaluator(TR::Node *node, TR::Code
...
+    }
+    else if (resultReg->getKind() == TR_VMR &&
+            boolNode->getOpCode().isLoadConst() &&
+            boolNode->getInt() != 0 &&
+            (numLanes == 8 && supportsOpMaskB) || (numLanes == 16) || (numLanes > 16 && supportsOpMaskDQ)
+    )
+    {
+        TR::InstOpCode maskOpcode = TR::InstOpCode::bad;
+
+        if (boolNode->getInt() != 0)
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRTreeEvaluator.cpp]: This condition is already checked in the `else if` L6072

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 2]
```diff
@@ -6030,9 +6030,160 @@ TR::Register *OMR::X86::TreeEvaluator::mstoreiEvaluator(TR::Node *node, TR::Code
...
+    );
+
+    bool supportsOpMaskDQ = cg->comp()->target().cpu.supportsFeature(OMR_FEATURE_X86_AVX512BW);
+    bool supportsOpMaskB = cg->comp()->target().cpu.supportsFeature(OMR_FEATURE_X86_AVX512DQ);
+
+#ifndef TR_TARGET_64BIT
+    TR_ASSERT_FATAL(false, "msplatsEvaluator is only supported on 64-bit");
+#endif
+
+    if (resultReg->getKind() == TR_VMR && boolNode->getOpCode().isLoadConst() && boolNode->getInt() == 0)
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRTreeEvaluator.cpp]: There are a lot of redundant checks in these first three conditions.   I think it can be re-written more succinctly as:
```
if (resultReg->getKind() == TR_VMR && boolNode->getOpCode().isLoadConst()) {
   if (boolNode->getInt() == 0) {
      ...
   } else if ((numLanes == 8 && supportsOpMaskB) || (numLanes == 16) || (numLanes > 16 && supportsOpMaskDQ)) {
      ...
   } else {
      ...
} else if (resultReg->getKind() == TR_VRF && boolNode->getOpCode().isLoadConst() && boolNode->getInt() == 0)) {
   ...
} else {
   ...
}
```

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 3]
```diff
@@ -6030,9 +6030,160 @@ TR::Register *OMR::X86::TreeEvaluator::mstoreiEvaluator(TR::Node *node, TR::Code
     return TR::TreeEvaluator::maskStoreEvaluator(node, cg);
 }
 
-TR::Register *OMR::X86::TreeEvaluator::msplatsEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+static bool isMSplatsConstTrue(TR::Node *node)
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRTreeEvaluator.cpp]: This function appears unused.

File: compiler/x/codegen/OMRTreeEvaluator.cpp (hunk 2 above)
Comment (BradleyWood) [reply compiler/x/codegen/OMRTreeEvaluator.cpp]: I simplified and removed the lane size checks. They were present in case of a situation where the chip supports AVX-512F but not AVX-512BW, however at some point I decided we wouldn't support native masks in that situation. Few if any chips exist with that combination of ISA features. It is also not easy to test such a config.

File: compiler/x/codegen/OMRCodeGenerator.cpp [hunk 4]
```diff
@@ -909,7 +909,10 @@ void OMR::X86::CodeGenerator::deactivateDependentDiscardableRegisters(TR::Regist
 bool OMR::X86::CodeGenerator::supportsOpMaskRegisters()
 {
     return self()->comp()->getOption(TR_EnableOpMaskRegisters)
-        && self()->comp()->target().cpu.supportsFeature(OMR_FEATURE_X86_AVX512F);
+        && self()->comp()->target().cpu.supportsFeature(OMR_FEATURE_X86_AVX512F)
+        && self()->comp()->target().cpu.supportsFeature(OMR_FEATURE_X86_AVX512VL)
+        && self()->comp()->target().cpu.supportsFeature(OMR_FEATURE_X86_AVX512BW)
+        && self()->comp()->target().cpu.supportsFeature(OMR_FEATURE_X86_AVX512DQ);
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRCodeGenerator.cpp]: For pathlength's sake, can you cache the cpu fetch in a local and use that instead?
```
TR::CPU *cpu = &self()->comp()->target().cpu;

if ( ...  && cpu.supportsFeature(OMR_FEATURE_X86_AVX512F) ...
```
```

Comment (BradleyWood) [reply compiler/x/codegen/OMRCodeGenerator.cpp]: [Done](https://github.com/eclipse-omr/omr/compare/1f113813d67e6341040fa3bfeadf282063c23a78..9c7ba8fc600028a5452a383568773f3b78cf9714)

File: compiler/x/codegen/OMRCodeGenerator.cpp [hunk 5]
```diff
@@ -908,8 +908,11 @@ void OMR::X86::CodeGenerator::deactivateDependentDiscardableRegisters(TR::Regist
 
 bool OMR::X86::CodeGenerator::supportsOpMaskRegisters()
 {
-    return self()->comp()->getOption(TR_EnableOpMaskRegisters)
-        && self()->comp()->target().cpu.supportsFeature(OMR_FEATURE_X86_AVX512F);
+    TR::CPU *cpu = &self()->comp()->target().cpu;
+
+    return self()->comp()->getOption(TR_EnableOpMaskRegisters) && cpu->supportsFeature(OMR_FEATURE_X86_AVX512F)
+        && cpu->supportsFeature(OMR_FEATURE_X86_AVX512VL) && cpu->supportsFeature(OMR_FEATURE_X86_AVX512BW)
+        && cpu->supportsFeature(OMR_FEATURE_X86_AVX512DQ);
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRCodeGenerator.cpp:915]: Apologies, I should have asked this earlier.  Why are all these AVX512 features now required in order to support mask registers?

Comment (BradleyWood) [reply compiler/x/codegen/OMRCodeGenerator.cpp:915]: Its not strictly required to support up to 16 lanes with native masking, but is required to support more than 16 lanes. I don't think there is many (if any) available hardware that supports AVX-512 but not these 4 ISA extensions, and the work to support an edge case for which we can't effectively test is not worth the effort.

Comment (0xdaryl) [reply compiler/x/codegen/OMRCodeGenerator.cpp:915]: OK, thanks for the explanation.

**Comment by 0xdaryl** [comment]
Jenkins build xlinux,win,x32linux,osx

**Review by 0xdaryl — APPROVED** [review]
